src/js.cookie.js

```diff
@@ -71,64 +60,72 @@
 			} catch(e) {}
 		}
 
-		return isFunction(converter) ? converter(value) : value;
+		return converter ? converter(value) : value;
 	}
 
 	function extend () {
-		var key, options;
 		var i = 0;
 		var result = {};
 		for (; i < arguments.length; i++) {
-			options = arguments[ i ];
-			for (key in options) {
+			var options = arguments[ i ];
+			for (var key in options) {
 				result[key] = options[key];
 			}
 		}
 		return result;
 	}
 
-	function isFunction (obj) {
-		return Object.prototype.toString.call(obj) === '[object Function]';
-	}
-
 	var api = function (key, value, options) {
-		var converter;
+		var converter, result;
+		var args = [].slice.call(arguments);
 
-		if (isFunction(value)) {
+		if (typeof value === 'function') {
 			converter = value;
-			value = undefined;
+			args.length = 1;
 		}
 
 		// Write
 
-		if (arguments.length > 1 && !converter) {
+		if (args.length > 1) {
 			options = extend(api.defaults, options);
 
 			if (typeof options.expires === 'number') {
-				var days = options.expires, t = options.expires = new Date();
-				t.setMilliseconds(t.getMilliseconds() + days * 864e+5);
+				var expires = new Date();
+				expires.setMilliseconds(expires.getMilliseconds() + options.expires * 864e+5);
+				options.expires = expires;
 			}
 
+			try {
+				result = JSON.stringify(value);
+				if (/^(?:\{[\w\W]*\}|\[[\w\W]*\])$/.test(result)) {
+					value = result;
+				}
+			} catch(e) {}
+
+			value = encode(String(value), unallowedCharsInValue);
+
 			return (document.cookie = [
-				encode(key, unallowedCharsInName), '=', processWrite(value),
-				options.expires ? '; expires=' + options.expires.toUTCString() : '', // use expires attribute, max-age is not supported by IE
-				options.path    ? '; path=' + options.path : '',
-				options.domain  ? '; domain=' + options.domain : '',
-				options.secure  ? '; secure' : ''
+				encode(key, unallowedCharsInName), '=', value,
+				options.expires && '; expires=' + options.expires.toUTCString(), // use expires attribute, max-age is not supported by IE
+				options.path    && '; path=' + options.path,
+				options.domain  && '; domain=' + options.domain,
+				options.secure  && '; secure'
 			].join(''));
 		}
 
 		// Read
 
-		var result = key ? undefined : {},
-			// To prevent the for loop in the first place assign an empty array
-			// in case there are no cookies at all. Also prevents odd result when
-			// calling "get()".
-			cookies = document.cookie ? document.cookie.split('; ') : [],
-			i = 0,
-			l = cookies.length;
+		if (!key) {
+			result = {};
+		}
+
+		// To prevent the for loop in the first place assign an empty array
+		// in case there are no cookies at all. Also prevents odd result when
+		// calling "get()"
+		var cookies = document.cookie ? document.cookie.split('; ') : [];
+		var i = 0;
 
-		for (; i < l; i++) {
+		for (; i < cookies.length; i++) {
```

Comment:
Well, we are testing in IE8+, but I am still not trying to use features that break IE7-, like [this line](https://github.com/js-cookie/js-cookie/pull/14/files#diff-b686f7b30e7b0c4f5ed6cd2e87241704R50) which uses `charAt()` instead of array notation.

If you think it's ok to drop IE 7- then we can rewrite that and save even more bytes.